<a href="https://colab.research.google.com/github/anasayago/PracticasAI/blob/main/TelecomX_LATAM_AnaSayago.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Diccionario de datos**
* customerID: número de identificación único de cada cliente
* Churn: si el cliente dejó o no la empresa

* gender: género (masculino y femenino)
* SeniorCitizen: información sobre si un cliente tiene o no una edad igual o mayor a 65 años
* Partner: si el cliente tiene o no una pareja
* Dependents: si el cliente tiene o no dependientes
* tenure: meses de contrato del cliente

* PhoneService: suscripción al servicio telefónico
* MultipleLines: suscripción a más de una línea telefónica

* InternetService: suscripción a un proveedor de internet
* OnlineSecurity: suscripción adicional de seguridad en línea
* OnlineBackup: suscripción adicional de respaldo en línea
* DeviceProtection: suscripción adicional de protección del dispositivo
* TechSupport: suscripción adicional de soporte técnico, menor tiempo de espera
* StreamingTV: suscripción de televisión por cable
* StreamingMovies: suscripción de streaming de películas

* Contract: tipo de contrato
* PaperlessBilling: si el cliente prefiere recibir la factura en línea
* PaymentMethod: forma de pago

* Charges.Monthly: total de todos los servicios del cliente por mes
* Charges.Total: total gastado por el cliente

#📌 Extracción

In [129]:
#cargar los datos desde la API de GIT
import pandas as pd
import numpy as np
df = pd.read_json('https://raw.githubusercontent.com/sthemonica/alura-voz/main/Dados/Telco-Customer-Churn.json')
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

#🔧 Transformación

In [130]:

customer_df = pd.json_normalize(df['customer'])
phone_df = pd.json_normalize(df['phone'])
internet_df = pd.json_normalize(df['internet'])
account_df = pd.json_normalize(df['account'])
#Concatenar el DataFrame original con los nuevos DataFrames
df = pd.concat([df.drop(columns=['customer', 'phone', 'internet', 'account']),
                customer_df,
                phone_df,
                internet_df,
                account_df], axis=1)
#Mostrar el DataFrame resultante
df.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

In [131]:
df.info()
colum_string = ['customerID','Churn','gender', 'Partner', 'Dependents', 'PhoneService',
                'MultipleLines', 'InternetService', 'OnlineSecurity',
                'OnlineBackup', 'DeviceProtection', 'TechSupport',
                'StreamingTV', 'StreamingMovies', 'Contract','PaperlessBilling',
                'PaymentMethod']
column_bool = ['SeniorCitizen']


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


In [132]:
#Convertir las columnas de tipo string
df[colum_string] = df[colum_string].astype(str)
#Convertir las columnas de tipo booleano
df[column_bool] = df[column_bool].astype(bool)
df['Charges.Total'] = pd.to_numeric(df['Charges.Total'], errors='coerce')
df['Charges.Total'] = df['Charges.Total'].fillna(0)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   bool   
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


In [133]:
#valores nulos
df.isnull().sum()
#valores unicos
df.nunique()

customerID          7267
Churn                  3
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
Charges.Monthly     1585
Charges.Total       6531
dtype: int64

In [134]:
#valores unicos de las columas de la lista colum_string
for column in colum_string:
    print({column})
    print(df[column].unique())
    print("\n")

{'customerID'}
['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']


{'Churn'}
['No' 'Yes' '']


{'gender'}
['Female' 'Male']


{'Partner'}
['Yes' 'No']


{'Dependents'}
['Yes' 'No']


{'PhoneService'}
['Yes' 'No']


{'MultipleLines'}
['No' 'Yes' 'No phone service']


{'InternetService'}
['DSL' 'Fiber optic' 'No']


{'OnlineSecurity'}
['No' 'Yes' 'No internet service']


{'OnlineBackup'}
['Yes' 'No' 'No internet service']


{'DeviceProtection'}
['No' 'Yes' 'No internet service']


{'TechSupport'}
['Yes' 'No' 'No internet service']


{'StreamingTV'}
['Yes' 'No' 'No internet service']


{'StreamingMovies'}
['No' 'Yes' 'No internet service']


{'Contract'}
['One year' 'Month-to-month' 'Two year']


{'PaperlessBilling'}
['Yes' 'No']


{'PaymentMethod'}
['Mailed check' 'Electronic check' 'Credit card (automatic)'
 'Bank transfer (automatic)']




In [135]:
#valores repetidos en customerID
df['customerID'].duplicated().sum()

np.int64(0)

In [136]:
#extraer de Churn los datos con valor '' guardarlos  y eliminarlos del DataFrame

#Eliminar las filas donde Churn es NaN  
registros_eliminados = df[df['Churn'] == '']
# Guardar los registros eliminados en un nuevo DataFrame
df = df[df['Churn'] != '']

#valores unicos de Churn
print(df['Churn'].unique())
registros_eliminados.head()

['No' 'Yes']


customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
30   0047-ZHDTW        Female          False      No         No      11   
75   0120-YZLQA          Male          False      No         No      71   
96   0154-QYHJU          Male          False      No         No      29   
98   0162-RZGMZ        Female           True      No         No       5   
175  0274-VVQOQ          Male           True     Yes         No      65   

    PhoneService MultipleLines InternetService  ...         OnlineBackup  \
30           Yes           Yes     Fiber optic  ...                   No   
75           Yes            No              No  ...  No internet service   
96           Yes            No             DSL  ...                  Yes   
98           Yes            No             DSL  ...                  Yes   
175          Yes           Yes     Fiber optic  ...                  Yes   

        DeviceProtection          TechSupport          StreamingTV  \
30                    No                   No                   No   
75   No internet service  No internet service  No internet service   
96                    No                  Yes                   No   
98                    No                  Yes                   No   
175                  Yes                   No                  Yes   

         StreamingMovies        Contract PaperlessBilling  \
30                    No  Month-to-month              Yes   
75   No internet service        Two year              Yes   
96                    No        One year              Yes   
98                    No  Month-to-month               No   
175                  Yes        One year              Yes   

                 PaymentMethod Charges.Monthly  Charges.Total  
30   Bank transfer (automatic)           79.00         929.30  
75     Credit card (automatic)           19.90        1355.10  
96            Electronic check           58.75        1696.20  
98     Credit card (automatic)           59.90         287.85  
175  Bank transfer (automatic)          103.15        6792.45  

[5 rows x 21 columns]

In [137]:
#Tranformar las columas YES y NO a 1 y 0 en varias columnas
df = df.applymap(lambda x: 1 if x == 'Yes' else 0 if x in ['No', 'No internet service','No phone service'] else x)


/var/folders/7l/3zvr6nj91ds_dl_w2h_7drsh0000gn/T/ipykernel_17821/2165195973.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df = df.applymap(lambda x: 1 if x == 'Yes' else 0 if x in ['No', 'No internet service','No phone service'] else x)


In [138]:
#agregar la columna nueva 'Charges.Diary' al DataFrame
df['Charges.Diary'] = (df['Charges.Monthly'] / 30).round(2)

In [139]:
new_column_names = {
    'customerID': 'ID Cliente',
    'Churn': 'Abandono',  
    'gender': 'Género',
    'SeniorCitizen': 'Mayor de 65 años',
    'Partner': 'Tiene Pareja',
    'Dependents': 'Tiene Dependientes',
    'tenure': 'Duración del Contrato (meses)',
    'PhoneService': 'Servicio Telefónico',
    'MultipleLines': 'Múltiples Líneas',
    'InternetService': 'Servicio de Internet',
    'OnlineSecurity': 'Seguridad en Línea',
    'OnlineBackup': 'Respaldo en Línea',
    'DeviceProtection': 'Protección del Dispositivo',
    'TechSupport': 'Soporte Técnico',
    'StreamingTV': 'TV por Cable',
    'StreamingMovies': 'Streaming de Películas',
    'Contract': 'Tipo de Contrato',
    'PaperlessBilling': 'Facturación Sin Papel',
    'PaymentMethod': 'Método de Pago',
    'Charges.Monthly': 'Costo Mensual',
    'Charges.Total': 'Costo Total',
    'Charges.Diary': 'Costo Diario'
}

# Renombramos las columnas
df.rename(columns=new_column_names, inplace=True)

In [140]:
# Diccionarios de entrada y reemplazo
replacement_dict = {
    'Género': {'Female': 'F', 'Male': 'M'},
    'Servicio de Internet': {'Fiber optic': 'Fibra Óptica'},
    'Tipo de Contrato':{'One year': 'Anual' ,'Month-to-month': 'Mensual' ,'Two year': 'Bianual'},
    'Método de Pago':{'Mailed check':'Cheque por Correo', 'Electronic check':'Cheque electrónico' ,'Credit card (automatic)':'Tarjeta de crédito (automático)','Bank transfer (automatic)': 'Transferencia bancaria (automático)'}  
}

# Reemplazar datos en el DataFrame usando los diccionarios
for column, replacements in replacement_dict.items():
    df[column] = df[column].replace(replacements)

df.head()

ID Cliente  Abandono Género  Mayor de 65 años  Tiene Pareja  \
0  0002-ORFBO         0      F             False             1   
1  0003-MKNFE         0      M             False             0   
2  0004-TLHLJ         1      M             False             0   
3  0011-IGKFF         1      M              True             1   
4  0013-EXCHZ         1      F              True             1   

   Tiene Dependientes  Duración del Contrato (meses)  Servicio Telefónico  \
0                   1                              9                    1   
1                   0                              9                    1   
2                   0                              4                    1   
3                   0                             13                    1   
4                   0                              3                    1   

   Múltiples Líneas Servicio de Internet  ...  Protección del Dispositivo  \
0                 0                  DSL  ...                           0   
1                 1                  DSL  ...                           0   
2                 0         Fibra Óptica  ...                           1   
3                 0         Fibra Óptica  ...                           1   
4                 0         Fibra Óptica  ...                           0   

   Soporte Técnico  TV por Cable  Streaming de Películas  Tipo de Contrato  \
0                1             1                       0             Anual   
1                0             0                       1           Mensual   
2                0             0                       0           Mensual   
3                0             1                       1           Mensual   
4                1             1                       0           Mensual   

   Facturación Sin Papel      Método de Pago  Costo Mensual Costo Total  \
0                      1   Cheque por Correo           65.6      593.30   
1                      0   Cheque por Correo           59.9      542.40   
2                      1  Cheque electrónico           73.9      280.85   
3                      1  Cheque electrónico           98.0     1237.85   
4                      1   Cheque por Correo           83.9      267.40   

   Costo Diario  
0          2.19  
1          2.00  
2          2.46  
3          3.27  
4          2.80  

[5 rows x 22 columns]

#📊 Carga y análisis

In [141]:
#Analisis descriptivo
df.describe(include='all')

ID Cliente     Abandono Género Mayor de 65 años  Tiene Pareja  \
count         7043  7043.000000   7043             7043   7043.000000   
unique        7043          NaN      2                2           NaN   
top     0002-ORFBO          NaN      M            False           NaN   
freq             1          NaN   3555             5901           NaN   
mean           NaN     0.265370    NaN              NaN      0.483033   
std            NaN     0.441561    NaN              NaN      0.499748   
min            NaN     0.000000    NaN              NaN      0.000000   
25%            NaN     0.000000    NaN              NaN      0.000000   
50%            NaN     0.000000    NaN              NaN      0.000000   
75%            NaN     1.000000    NaN              NaN      1.000000   
max            NaN     1.000000    NaN              NaN      1.000000   

        Tiene Dependientes  Duración del Contrato (meses)  \
count          7043.000000                    7043.000000   
unique                 NaN                            NaN   
top                    NaN                            NaN   
freq                   NaN                            NaN   
mean              0.299588                      32.371149   
std               0.458110                      24.559481   
min               0.000000                       0.000000   
25%               0.000000                       9.000000   
50%               0.000000                      29.000000   
75%               1.000000                      55.000000   
max               1.000000                      72.000000   

        Servicio Telefónico  Múltiples Líneas Servicio de Internet  ...  \
count           7043.000000       7043.000000                 7043  ...   
unique                  NaN               NaN                    3  ...   
top                     NaN               NaN         Fibra Óptica  ...   
freq                    NaN               NaN                 3096  ...   
mean               0.903166          0.421837                  NaN  ...   
std                0.295752          0.493888                  NaN  ...   
min                0.000000          0.000000                  NaN  ...   
25%                1.000000          0.000000                  NaN  ...   
50%                1.000000          0.000000                  NaN  ...   
75%                1.000000          1.000000                  NaN  ...   
max                1.000000          1.000000                  NaN  ...   

        Protección del Dispositivo  Soporte Técnico  TV por Cable  \
count                  7043.000000      7043.000000   7043.000000   
unique                         NaN              NaN           NaN   
top                            NaN              NaN           NaN   
freq                           NaN              NaN           NaN   
mean                      0.343888         0.290217      0.384353   
std                       0.475038         0.453895      0.486477   
min                       0.000000         0.000000      0.000000   
25%                       0.000000         0.000000      0.000000   
50%                       0.000000         0.000000      0.000000   
75%                       1.000000         1.000000      1.000000   
max                       1.000000         1.000000      1.000000   

        Streaming de Películas  Tipo de Contrato  Facturación Sin Papel  \
count              7043.000000              7043            7043.000000   
unique                     NaN                 3                    NaN   
top                        NaN           Mensual                    NaN   
freq                       NaN              3875                    NaN   
mean                  0.387903               NaN               0.592219   
std                   0.487307               NaN               0.491457   
min                   0.000000               NaN               0.000000   
25%                   0.000000               NaN               0.000

### Tasa de abandono
La tasa de abandono promedio es del 26.54%. Esto indica que aproximadamente un cuarto de los clientes deja el servicio, lo cual es una señal de alerta para la empresa.
### Demografía del cliente:
* Género: La mayoría son hombres (51.69%).
* Mayor de 65 años: Solo el 48.31% de los clientes son mayores de 65 años, lo que sugiere que la mayoría son más jóvenes.
* Tiene Pareja: El 48.30%  tienen pareja.
* Tiene Dependientes: El 29.96% tienen dependientes.

### Duración del contrato:
Duración promedio 32.37 meses, con una desviación estándar de 24.56 meses. 

### Servicios adicionales:
* Servicio telefónico: El 90.32% tienen servicio telefónico.
* Múltiples líneas: El 42.18% tienen múltiples líneas.
* Servicio de internet: El 90.32% tienen servicio de internet.
* Protección del dispositivo: El 34.39% tienen protección del dispositivo.
* Soporte técnico: El 29.02% tienen soporte técnico.
* TV por cable: El 38.44% tienen TV por cable.
* Streaming de películas: El 38.79% tienen streaming de películas.

### Tipo de contrato y facturación:
* Tipo de Contrato: La mayoría son mensuales (59.22%).
* Facturación Sin Papel: El 59.22% prefieren la facturación sin papel.
* Método de Pago: El 23.65% prefieren el cheque electrónico.

### Costos:
* Costo mensual promedio de $64.76
* Costo total promedio es de $2,279.73
* Costo diario promedio es de $2.16

### Conclusión

* Retención de clientes: La alta tasa de abandono (26.54%) es una preocupación significativa. La empresa debe investigar las razones detrás de este abandono y desarrollar estrategias para mejorar la retención de clientes, como ofrecer mejores servicios al cliente, promociones de fidelización, o mejorar la calidad del servicio.
* Demografía del cliente: La mayoría de los clientes son hombres y no mayores de 65 años. Esto sugiere que la empresa debe enfocarse en segmentos de mercado más jóvenes y masculinos para sus campañas de marketing y desarrollo de productos.
* Servicios Adicionales: La alta adopción de servicios como el servicio telefónico y el servicio de internet indica que estos son servicios clave. Sin embargo, la baja adopción de servicios adicionales como la protección del dispositivo y el soporte técnico sugiere una oportunidad para aumentar los ingresos mediante la promoción de estos servicios.
* Duración del Contrato: La variabilidad en la duración del contrato sugiere que la empresa debe trabajar en la estandarización y la mejora de la experiencia del cliente para asegurar contratos más largos y estables.
* Costos: Los costos mensuales y totales son relativamente bajos, lo que puede ser una ventaja competitiva. Sin embargo, la empresa debe asegurarse de que estos costos sean sostenibles y que no estén afectando la calidad del servicio.


#📄Informe final